### PR TITLE
[Dev] Add more tests for LayerwiseDistOpt with dist_ckpt

### PR DESCRIPTION
**Pull request by BoxiangW**
# What does this PR do ?
[PR to main](https://github.com/NVIDIA/Megatron-LM/pull/2241)

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by dimapihtar** [comment]
@BoxiangW looks like we need to add "functional_tests" label in order to trigger that functional tests you added in this PR.

**Review by dimapihtar — APPROVED** [review]
LGTM. Thank you!

File: tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,621 @@
...
+                    tp=tp_pp[0],
+                    pp=tp_pp[1],
+                    bf16=bf16,
+                    dist_opt=False,
+                    initialize_fn=initialize_gpt_model,
+                    optimizer='dist_muon',
+                )
+
+                # Save checkpoint A
+                model_sharded_sd_A = model_A[0].sharded_state_dict()
```
Comment (sbak5) [inline tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py:215]: Can you use `generate_state_dict` at `training/checkpointing.py` instead? 
You can pass arguments to `False` or `none` if you want to exclude rng or opt_param_scheduler
This will make some `non-shardable` objects, which will be saved in `common.pt`

Comment (sbak5) [reply tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py:215]: I think the test below for common_state_dict serves this purpose

Comment (BoxiangW) [reply tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py:215]: Then we can resolve this conversation?

File: megatron/core/optimizer/optimizer_config.py [hunk 2]
```diff
@@ -146,8 +146,8 @@ class OptimizerConfig:
     muon_tp_mode: str = "blockwise"
     """How to perform NS calculation for tensor parallel weights. Defaults to "blockwise"."""
 
-    muon_extra_scale_factor: float = 1.0
-    """Additional scale factor for the muon update."""
+    muon_extra_scale_factor: float = 0.2
```
Comment (skyw) [inline megatron/core/optimizer/optimizer_config.py]: FIX: Revert this change. Don't change this default scale factor.
ESPECIALLY not in a unrelated PR.

File: megatron/core/dist_checkpointing/dict_utils.py [hunk 3]
```diff
@@ -137,30 +137,30 @@ def diff(x1: Any, x2: Any, prefix: Tuple = ()) -> Tuple[list, list, list]:
 
 def inspect_types(x: Any, prefix: Tuple = (), indent: int = 4):
     """Helper to print types of (nested) dict values."""
-    print_indent = lambda: print(" " * indent * len(prefix), end="")
+    print_indent = lambda: print(" " * indent * len(prefix), end="")  # pylint: disable=bad-builtin
```
Comment (skyw) [inline megatron/core/dist_checkpointing/dict_utils.py]: too many inline disable. just disable it for the section.
```python
# pylint: disable=bad-builtin
...
# pylint: enable=bad-builtin
```

File: megatron/training/arguments.py [hunk 4]
```diff
@@ -1970,7 +1970,7 @@ def _add_regularization_args(parser):
     group.add_argument('--muon-tp-mode', type=str, default='blockwise',
                        choices=['blockwise', 'duplicated', 'distributed'],
                        help='How to perform NS calculation for tensor model parallel weights')
-    group.add_argument('--muon-extra-scale-factor', type=float, default=1.0,
+    group.add_argument('--muon-extra-scale-factor', type=float, default=0.2,
```
Comment (skyw) [inline megatron/training/arguments.py]: FIX: Revert this change.

File: tests/test_utils/python_scripts/launch_nemo_run_workload.py [hunk 5]
```diff
@@ -133,10 +133,10 @@ def main(
         succeeded = str(job_dict["status"]) == "SUCCEEDED"
 
         if succeeded:
-            logger.info(f"Job succeeded with status: {job_dict["status"]}")
+            logger.info(f"Job succeeded with status: {job_dict['status']}")
```
Comment (skyw) [inline tests/test_utils/python_scripts/launch_nemo_run_workload.py]: revert meaningless format change, unless their is a new style guide?

File: tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py [hunk 6]
```diff
@@ -0,0 +1,621 @@
...
+        # Check if optimizer is ChainedOptimizer (expected for standard setup)
+        if isinstance(optimizer, ChainedOptimizer):
+            total_params = sum(
+                len(group['params'])
+                for opt in optimizer.chained_optimizers
+                for group in opt.param_groups
+            )
+            assert total_params > 0, "No parameters found in optimizer"
+
+    @pytest.mark.parametrize('tp_pp', [(1, 2), (2, 1), (2, 2)])
```
Comment (skyw) [inline tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py]: Fix: Bad variable grouping. should use TP and PP separately.
Same for all other tests.

File: tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py [hunk 7]
```diff
@@ -0,0 +1,621 @@
...
+            # Load should work for same TP/PP or handle differences gracefully
+            if src_tp_pp == dest_tp_pp:
+                state_dict = load(load_sharded_sd, ckpt_dir)
+                optimizer_B.load_state_dict(state_dict)
+            else:
+                # For different TP/PP, load may succeed or fail depending on compatibility
+                try:
+                    state_dict = load(load_sharded_sd, ckpt_dir)
+                    optimizer_B.load_state_dict(state_dict)
+                except Exception:
```
Comment (skyw) [inline tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py]: Don't silently pass tests. Can skip the test if there is a known reason.

File: tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py [hunk 8]
```diff
@@ -0,0 +1,621 @@
...
+
+            # Create new optimizer and load
+            model_new, optimizer_new = setup_moe_model_and_optimizer(
+                seed=3, tp=tp, pp=pp, ep=ep, bf16=True, dist_opt=False, optimizer='dist_muon'
+            )
+
+            model_sharded_sd = model_new[0].sharded_state_dict()
+            load_sharded_sd = optimizer_new.sharded_state_dict(model_sharded_sd, is_loading=True)
+            state_dict = load(load_sharded_sd, ckpt_dir)
+            optimizer_new.load_state_dict(state_dict)
```
Comment (skyw) [inline tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py:408]: Need to check loaded optimizer matches reference.

File: tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py [hunk 9]
```diff
@@ -0,0 +1,621 @@
...
+                        optimizer=optimizer_type,
+                    )
+                # Load optimizer B from checkpoint
+                load_checkpoint_no_arg_checks(model, optimizer_B, None)
+
+                # Get optimizer B param state
+                optim_param_state_B = optimizer_B.state_dict()
+
+                # Test both param state dicts are equal
+                diffs = diff(optim_param_state_A, optim_param_state_B)
```
Comment (skyw) [inline tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py]: Try not to use self made diff function as testing metric, it mixes things together unnecessarily. e.g. a bug in diff could have very large blast radius because all tests depend on it become faulty.
Better to explicitly compare everything here.

**Review by skyw — CHANGES_REQUESTED** [review]
This PR should not be merged.
* The title says it is only adding tests, but there a lot of code change including public API change. They should be spereate.
* Test is not thorough, i.e. it doesn't catch a know issue that load fail on empty state_dict.
* Plus some other improvement suggested.

File: tests/test_utils/python_scripts/launch_nemo_run_workload.py (hunk 5 above)
Comment (BoxiangW) [reply tests/test_utils/python_scripts/launch_nemo_run_workload.py]: Yes, it was a pylint guide

Comment (skyw) [reply tests/test_utils/python_scripts/launch_nemo_run_workload.py]: but lines above and below all use double quote?

File: megatron/core/dist_checkpointing/dict_utils.py (hunk 3 above)
Comment (FDecaYed) [reply megatron/core/dist_checkpointing/dict_utils.py]: I'd suggest we just remove changes from this file(the only change is a comment). and thus we don't need to fix lint for it.
For that comment, I think it is ok for it to catch our empty list issue and we should fix it from layerwise side

File: megatron/core/optimizer/optimizer_config.py (hunk 2 above)
Comment (BoxiangW) [reply megatron/core/optimizer/optimizer_config.py]: Reverted, this was recommended by Mkail and we can change it in a separate PR then

File: tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py (hunk 8 above)
Comment (BoxiangW) [reply tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py:408]: It is tested in other tests below

File: tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py (hunk 6 above)
Comment (BoxiangW) [reply tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py]: This is not applicable for all the cases, since we have 8GPU max, if we do
```
    @pytest.mark.parametrize('tp', [2, 4])
    @pytest.mark.parametrize('pp', [2, 4])
```
It will reach 16 GPUs in total

File: tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py (hunk 7 above)
Comment (BoxiangW) [reply tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py]: Removed this special treatment

File: tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py (hunk 9 above)
Comment (BoxiangW) [reply tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py]: Added a similar but simpler explicitly checking in this file as a util function

File: tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py (hunk 6 above)
Comment (skyw) [reply tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py]: you can skip test (with a message) if total number of GPU exceeded.

File: tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py (hunk 8 above)
Comment (skyw) [reply tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py:408]: I see.
Common suggestion is to state what being tested in test name, e.g. test_blah_match_oops

File: tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py [hunk 10]
```diff
@@ -0,0 +1,627 @@
...
+        """Test LayerWiseDistributedOptimizer with different DP sizes."""
+        # Use PP to vary DP size while keeping world size constant
+        world_size = 8
+        if world_size % dp_size != 0:
+            pytest.skip(f"World size {world_size} not divisible by DP size {dp_size}")
+
+        pp = dp_size
+        tp = world_size // pp
+
+        if tp == 0:
```
Comment (skyw) [inline tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py:449]: Should check broadly whether world_size is divisible by pp.

Comment (BoxiangW) [reply tests/unit_tests/dist_checkpointing/test_layer_wise_optimizer.py:449]: Fixed this part

**Comment by Victarry** [comment]
Hi @BoxiangW , the golden values of this MR is not added, which cause the functional tests fail https://gitlab-master.nvidia.com/ADLR/megatron-lm/-/pipelines/38398052.
